### ISSUE TITLE
Probe failed to find any functional HID backend on Raspberry Pi

### ISSUE THREAD

**Issue by thomasvoorend** [abcminiuser/python-elgato-streamdeck#32]
We've tried to use HIDAPI for our streamdeck on the Raspberry Pi. With HIDAPI, the streamdeck works normal for a couple of minutes. After some time, the software do not receive any key input.

Now we were trying to remove the  HIDAPI python package and use HID, but it throws the following error on starting our program:

`Traceback (most recent call last):
  File "streamdeck.py", line 180, in <module>
    main()
  File "streamdeck.py", line 162, in main
    streamdeck()
  File "streamdeck.py", line 139, in streamdeck
    streamdecks = DeviceManager().enumerate()
  File "/home/pi/.local/lib/python3.7/site-packages/StreamDeck/DeviceManager.py", line 90, in __init__
    self.transport = self._get_transport(transport)
  File "/home/pi/.local/lib/python3.7/site-packages/StreamDeck/DeviceManager.py", line 82, in _get_transport
    raise ProbeError("Probe failed to find any functional HID backend.", probe_errors)
StreamDeck.DeviceManager.ProbeError: ('Probe failed to find any functional HID backend.', {'hid': ImportError('Unable to load any of the following libraries:libhidapi-hidraw.so libhidapi-hidraw.so.0 libhidapi-libusb.so libhidapi-libusb.so.0 libhidapi-iohidmanager.so libhidapi-iohidmanager.so.0 libhidapi.dylib hidapi.dll libhidapi-0.dll'), 'hidapi': ImportError('Unable to load any of the foll`

Has someone experience with this?

**Comment by abcminiuser** [comment]
Try the installing the `libudev-dev and `libusb-1.0-0-dev` packages for your distro - it should install the libusb hidapi backend required by the HID python library.

**Comment by thomasvoorend** [comment]
I've tried it, but unfortunately without luck. 

After that I have tried to remove hid (and just for sure hidapi) and reinstall it all again. But I still get the same error message.
Do you have any idea if anything else could be the cause of this error?
```
pi@raspberrypi:~ $ sudo pip3 uninstall hid hidapi
Uninstalling hid-1.0.4:
  Would remove:
    /usr/local/lib/python3.7/dist-packages/hid-1.0.4.dist-info/*
    /usr/local/lib/python3.7/dist-packages/hid/*
Proceed (y/n)? y
  Successfully uninstalled hid-1.0.4
Skipping hidapi as it is not installed.
pi@raspberrypi:~ $ sudo apt install -y libudev-dev libusb-1.0-0-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libudev-dev is already the newest version (241-7~deb10u2+rpi1).
libusb-1.0-0-dev is already the newest version (2:1.0.22-2).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
pi@raspberrypi:~ $ sudo apt install -y libjpeg-dev zlib1g-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libjpeg-dev is already the newest version (1:1.5.2-2).
zlib1g-dev is already the newest version (1:1.2.11.dfsg-1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
pi@raspberrypi:~ $ pip3 install hid
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting hid
  Using cached https://www.piwheels.org/simple/hid/hid-1.0.4-py3-none-any.whl
Installing collected packages: hid
Successfully installed hid-1.0.4
pi@raspberrypi:~ $ pip3 install pillow
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: pillow in ./.local/lib/python3.7/site-packages (6.2.1)
pi@raspberrypi:~ $ sudo tee /etc/udev/rules.d/10-streamdeck.rules << EOF
> SUBSYSTEMS=="usb", ATTRS{idVendor}=="0fd9", GROUP="users"
> EOF
SUBSYSTEMS=="usb", ATTRS{idVendor}=="0fd9", GROUP="users"
pi@raspberrypi:~ $ pip3 install streamdeck
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: streamdeck in /usr/local/lib/python3.7/dist-packages (0.6.3)
pi@raspberrypi:~ $ python3 ~/python-elgato-streamdeck/src/example_deckinfo.py 
Traceback (most recent call last):
  File "/home/pi/python-elgato-streamdeck/src/example_deckinfo.py", line 44, in <module>
    streamdecks = DeviceManager().enumerate()
  File "/home/pi/python-elgato-streamdeck/src/StreamDeck/DeviceManager.py", line 90, in __init__
    self.transport = self._get_transport(transport)
  File "/home/pi/python-elgato-streamdeck/src/StreamDeck/DeviceManager.py", line 82, in _get_transport
    raise ProbeError("Probe failed to find any functional HID backend.", probe_errors)
StreamDeck.DeviceManager.ProbeError: ('Probe failed to find any functional HID backend.', {'hid': OSError('The libhidapi-hidraw0 backend is currently blacklisted for StreamDeck devices.'), 'hidapi': TypeError("enumerate() got an unexpected keyword argument 'vendor_id'")})
```

**Comment by thomasvoorend** [comment]
Another try on a clean install of Raspbian Buster, unfortunately still the same error.

**Comment by abcminiuser** [comment]
_(Repeating here what I just sent to you in email, in case someone else hits the same issue and finds this thread.)_

HID access via Python is currently a real mess. The TL;DR answer if you don't want to read the below is to try running a `sudo apt-get remove libhidapi-hidraw0` on the Raspberry Pi.

The longer answer is that there are two packages that I support for the HID communication backend to the StreamDeck: `hid` and `hidapi`. The `hid` library is the better of the two, as it just wraps the system installed HIDAPI library, which means you can keep the actual underlying OS code up to date just by making sure the system packages are up to date. The `hidapi` library is self-contained which makes it easier to install on Windows, but contains a whole bunch of bugs which can cause various strange issues if the library is used over a long period of time.

Further complicating matters is that the `hid` and `hidapi` packages conflict with each other, so you can only install one or the other. The library is probing for both and your log shows it found the `hid` library, which is good! The probe is failing however due to the last complication: there's two possible OS layers the `hid` package can use, and it will automatically pick the first one it finds.

The first OS layer is `libhidapi-hidraw`, which works decently but cannot talk to the original StreamDeck due to an API limitation. Second is `libhidapi-libusb`, which works for all the StreamDecks. The problem: the `hid` Python library will use `libhidapi-hidraw` over `libhidapi-libusb` if both are installed, which isn't what we want. Removing the `libhidapi-hidraw0` should force the `hid` library to find the only working OS layer, the one we want.

To summarize: the possible library/system package combinations are:

Works:
- **python-elgto-streamdeck** (python library) -> **hid** (python library) -> **libhidapi-libusb** (OS package)

Kinda works, but buggy:
- **python-elgto-streamdeck** (python library) -> **hidapi** (python library)

Doesn't work (blacklisted, as it won't talk to V1 StreamDecks):
- **python-elgto-streamdeck** (python library) -> **hid** (python library) -> **libhidapi-hidraw** (OS package)

**Comment by AndreKager** [comment]
Hello Dean, we got it working by doing the following:

sudo apt-get remove libhidapi-hidraw0
sudo apt-get install libhidapi-libusb0

Thanks for your help! Now debugging out code :-)

Cheers!

**Comment by trandersen-ufst** [comment]
> _(Repeating here what I just sent to you in email, in case someone else hits the same issue and finds this thread.)_

That was exactly what I did when playing with your examples.  Thank you for the thorough explanation.
